src/main/java/io/cdap/http/internal/HttpResourceHandler.java

```diff
@@ -277,7 +289,22 @@ public HttpMethodInfo getDestinationMethod(HttpRequest request, HttpResponder re
         if (!terminated) {
           // Wrap responder to make post hook calls.
           responder = new WrappedHttpResponder(responder, handlerHooks, request, info);
-          return httpResourceModel.handle(request, responder, matchedDestination.getGroupNameValues());
+          HttpMethodInfo methodInfo = httpResourceModel.handle(request, responder,
+                                                               matchedDestination.getGroupNameValues());
+          // check for authentication and authorization
+          if (authHandler != null) {
+            boolean shouldExecute = true;
```

Comment:
I think it is easier to understand by using early termination:

```java
if (authHandler != null) {
  if (method.isSecure() && !authHandler.isAuthenticated(request)) {
    throw new AuthHandlerException(...);
  }
  String[] requiredRoles = method.getRequiredRoles();
  if (requiredRoles != null && !authHandler.hasRoles(request, requiredRoles)) {
    throw new AuthHandlerException(...);
  }
}
```

Reply:
BTW, I think not authenticated vs not authorized should be two different exceptions.